### PR TITLE
Add --restart flag to pod create

### PR DESCRIPTION
**Pull request by umohnani8**
Add --restart flag to pod create to allow users to set the restart policy for the pod, which applies to all the containers in the pod. This reuses the restart policy already there for containers and has the same restart policy options. Add "never" to the restart policy options to match k8s syntax. It is a synonym for "no" and does the exact same thing where the containers are not restarted once exited.
Only the containers that have exited will be restarted based on the restart policy, running containers will not be restarted when an exited container is restarted in the same pod (same as is done in k8s).

Add Restarts column to the podman ps and podman pod ps output to show how many times a
container was restarted based on its restart policy when --format={{.Restarts}}.

Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
Add --restart flag to pod create to be able to set the restart policy for all the containers in a pod.
Add Restarts column to the output of podman ps to show the number of times a container has been restarted based on its restart policy when --format={{.Restarts}}.
Add Restarts column to the output of podman pod ps when --format={{.Restarts}} to show the total number of container restarts in a pod.
```

### CODE REVIEW

File: test/e2e/pod_create_test.go [hunk 1]
```diff
@@ -1213,4 +1214,131 @@ ENTRYPOINT ["sleep","99999"]
...
+		ps = podmanTest.Podman([]string{"ps", "-a", "--filter", "name=" + testCtr, "--format", "{{.Status}}"})
+		ps.WaitWithDefaultTimeout()
+		Expect(ps).Should(Exit(0))
+		Expect(ps.OutputToString()).To(ContainSubstring("Exited"))
+		ps = podmanTest.Podman([]string{"ps", "-a", "--filter", "name=" + runningCtr, "--format", "{{.Status}}"})
+		ps.WaitWithDefaultTimeout()
+		Expect(ps).Should(Exit(0))
+		Expect(ps.OutputToString()).To(ContainSubstring("Up"))
+
+		// Restart policy set to no
```
Comment (TomSweeneyRedHat) [inline test/e2e/pod_create_test.go:1296]: It would be good to test "never" too.

**Comment by TomSweeneyRedHat** [comment]
After a quick skim, looks good in general.  Test aren't at all happy.

File: cmd/podman/containers/create.go [hunk 2]
```diff
@@ -138,6 +138,10 @@ func create(cmd *cobra.Command, args []string) error {
 		cliVals.InitContainerType = initctr
 	}
 
+	if cmd.Flags().Changed("pod") && strings.HasPrefix(cmd.Flag("pod").Value.String(), "new:") && cmd.Flags().Changed("restart") {
+		return fmt.Errorf("must specify policy for --restart during pod creation; containers in a pod cannot have individual restart policies")
```
Comment (vrothberg) [inline cmd/podman/containers/create.go:142]: That is a too big hammer breaking backwards compatibility.  Existing workloads with individual restart policies of containers inside a pod must continue to work.  If we want to restrict it, the check should happen in the backend (probably specgen) to account for REST calls.

I didn't read the full PR yet but at the moment I do not understand the restriction.  The restart policy of a container may be a different one than for the entire pod.

File: cmd/podman/containers/ps.go [hunk 3]
```diff
@@ -199,12 +200,6 @@ func ps(cmd *cobra.Command, _ []string) error {
 	if err != nil {
 		return err
 	}
-	if len(listOpts.Sort) > 0 {
```
Comment (vrothberg) [inline cmd/podman/containers/ps.go:202]: Doesn't that break `podman ps --sort`?

File: cmd/podman/containers/ps.go [hunk 4]
```diff
@@ -300,6 +295,7 @@ func createPsOut() ([]map[string]string, string) {
 		"PIDNS":        "pidns",
 		"Pod":          "pod id",
 		"PodName":      "podname", // undo camelcase space break
+		"RestartCount": "restarts",
```
Comment (vrothberg) [inline cmd/podman/containers/ps.go:298]: I prefer this to be part of another commit.  Adding restarts to pods shouldn't require changes to `container ps`.

File: docs/source/markdown/podman-pod-create.1.md.in [hunk 5]
```diff
@@ -143,6 +143,11 @@ but only by the pod itself.
 
 @@option replace
 
+@@option restart
+
+The restart policy is set at the pod level so all containers in the pod have the same restart policy configured.
```
Comment (vrothberg) [inline docs/source/markdown/podman-pod-create.1.md.in:148]: I do not understand this restriction. The restart policy of a pod may very well be different than the one of its individual containers.

File: docs/source/markdown/podman-pod-create.1.md.in [hunk 6]
```diff
@@ -143,6 +143,11 @@ but only by the pod itself.
...
+When the restart policy is set to **no** or **never**, the exit-policy of the pod is set to **stop**. This will ensure that infra container also exits once all the containers in the pod exit.
```
Comment (vrothberg) [inline docs/source/markdown/podman-pod-create.1.md.in:149]: Why only on `no` or `never`?  What about pods with not infra container?  Is this information important to readers?

File: libpod/define/container.go [hunk 7]
```diff
@@ -7,6 +7,9 @@ const (
 	RestartPolicyNone = ""
 	// RestartPolicyNo is identical in function to RestartPolicyNone.
 	RestartPolicyNo = "no"
+	// RestartPolicyNever is the same as no, which indicates that the container
+	// should not be restarted when exited.
+	RestartPolicyNever = "never"
```
Comment (vrothberg) [inline libpod/define/container.go:12]: No need for a new const.  Let's just map the string "never" to `RestartPolicyNo` when parsing.

File: test/e2e/pod_create_test.go [hunk 8]
```diff
@@ -1213,4 +1214,131 @@ ENTRYPOINT ["sleep","99999"]
 		Expect(strings[0]).Should(ContainSubstring("size=10240k"))
 	})
 
+	FIt("podman pod create --restart", func() {
```
Comment (vrothberg) [inline test/e2e/pod_create_test.go:1217]:
```suggestion
	It("podman pod create --restart", func() {
```

Comment (vrothberg) [inline test/e2e/pod_create_test.go:1217]: Consider turning this into a table test.  It's very large and hence hard to read/maintain.

**Review by vrothberg — COMMENTED** [review]
I had something else in mind for `pod create --restart`.  Currently, it does not yield a `podman pod restart` but instead, it'll set it for all containers.  There are still some open questions in the design doc as well.

@mheon @giuseppe PTAL

File: cmd/podman/containers/ps.go [hunk 9]
```diff
@@ -309,7 +305,7 @@ func createPsOut() ([]map[string]string, string) {
 	if listOpts.Namespace {
 		row = "{{.ID}}\t{{.Names}}\t{{.Pid}}\t{{.Namespaces.Cgroup}}\t{{.Namespaces.IPC}}\t{{.Namespaces.MNT}}\t{{.Namespaces.NET}}\t{{.Namespaces.PIDNS}}\t{{.Namespaces.User}}\t{{.Namespaces.UTS}}"
 	} else {
-		row = "{{.ID}}\t{{.Image}}\t{{.Command}}\t{{.CreatedHuman}}\t{{.Status}}\t{{.Ports}}\t{{.Names}}"
+		row = "{{.ID}}\t{{.Image}}\t{{.Command}}\t{{.CreatedHuman}}\t{{.Status}}\t{{.RestartCount}}\t{{.Ports}}\t{{.Names}}"
```
Comment (Luap99) [inline cmd/podman/containers/ps.go:308]: IMO we should not change the default output, I see no reason why I would want to see  the restart count by default. Users can always supply their own format template. Also tihs breaks backwards compat and docker compat if users really on the current ordering in scripts, not a good idea regardless but I don't think this justifies breaking them.

**Comment by umohnani8** [comment]
If we are to follow how kubernetes does restarts for pods, it doesn't restart the whole pod based on the policy, the containers in the pod are what is restarted. So when we set the restart policy for a pod, we are setting the restart policy for all the containers in the pod and a container in the pod cannot have a different policy set for it. If the whole pod needs to be restarted, k8s removes the pod and creates a new one in place of it and that is a whole different thing that happens when using Deployments or Jobs or when the user deletes the pod and the policy was set to Always.
https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

The design doc outlined the object of --restart for pod to apply it to the containers and have the containers restart based on the policy and there were no objections on that. We can continue the discussions in the design doc if needed.

**Comment by vrothberg** [comment]
> If we are to follow how kubernetes does restarts for pods, it doesn't restart the whole pod based on the policy, the containers in the pod are what is restarted. So when we set the restart policy for a pod, we are setting the restart policy for all the containers in the pod and a container in the pod cannot have a different policy set for it. If the whole pod needs to be restarted, k8s removes the pod and creates a new one in place of it and that is a whole different thing that happens when using Deployments or Jobs or when the user deletes the pod and the policy was set to Always. https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

I _personally_ don't mind K8s in this context as it's outside of `kube play`.

> The design doc outlined the object of --restart for pod to apply it to the containers and have the containers restart based on the policy and there were no objections on that. We can continue the discussions in the design doc if needed.

OK, let's do that :+1:   I misinterpreted the part in the doc w.r.t "apply it to all containers".  If the pod gets restarted it affects all its containers in a way.

File: test/e2e/pod_create_test.go (hunk 1 above)
Comment (umohnani8) [reply test/e2e/pod_create_test.go:1296]: added a test for "never"

File: cmd/podman/containers/create.go (hunk 2 above)
Comment (umohnani8) [reply cmd/podman/containers/create.go:142]: Is this still considered breaking backwards compatibility? We agreed on following k8s and in that case the restart policy can only be set on the pod level and is applied to all the containers in the pod. So we cannot set --restart for containers in a pod.

File: cmd/podman/containers/ps.go (hunk 3 above)
Comment (umohnani8) [reply cmd/podman/containers/ps.go:202]: No, this code is repeated from https://github.com/containers/podman/pull/17627/files#diff-d1aefcf956b4e079a0684a5f5a5d545d4be20870f172fdac95c77cac1fd90482R179. I noticed it while updating ps but can remove it if it is preferred to do in another PR.

File: cmd/podman/containers/ps.go (hunk 4 above)
Comment (umohnani8) [reply cmd/podman/containers/ps.go]: Adding restarts to podman ps is part of the design doc goals. It also helps with testing the pod restart policy feature.

I separated out the commits for podman ps, podman pod ps, and podman pod create --restart.

File: docs/source/markdown/podman-pod-create.1.md.in (hunk 6 above)
Comment (umohnani8) [reply docs/source/markdown/podman-pod-create.1.md.in:149]: On `no` and `never`, the containers will never restart. Since exit-policy is used to stop the pod completely when the containers exit, I thought to use it for this use case as the containers will never be restarted and we can make the pod exit once they exit. I may have misunderstood the purpose of `--exit-policy`.

File: libpod/define/container.go (hunk 7 above)
Comment (umohnani8) [reply libpod/define/container.go]: fixed

File: test/e2e/pod_create_test.go (hunk 8 above)
Comment (umohnani8) [reply test/e2e/pod_create_test.go]: fixed

Comment (umohnani8) [reply test/e2e/pod_create_test.go]: Cleaned up the tests a bit and split them up.

File: libpod/options.go [hunk 10]
```diff
@@ -1998,6 +2000,43 @@ func WithPodExitPolicy(policy string) PodCreateOption {
...
+	return func(pod *Pod) error {
+		if pod.valid {
+			return define.ErrPodFinalized
+		}
+
+		switch policy {
+		case define.RestartPolicyNo, define.RestartPolicyOnFailure, define.RestartPolicyAlways, define.RestartPolicyUnlessStopped:
+			pod.config.RestartPolicy = policy
+		case define.RestartPolicyNone:
+			pod.config.RestartPolicy = define.RestartPolicyAlways
```
[the remark this answers is not in the document]
Comment (umohnani8) [reply libpod/options.go:2035]: None means that a restart policy wasn't set, in that case the default restart policy should be set to `Always`

File: cmd/podman/containers/ps.go (hunk 9 above)
Comment (umohnani8) [reply cmd/podman/containers/ps.go]: Added a --restart flag to display the restart count when set to true so we don't break backwards compatibility and users can use the flag or template to view the restart count for containers

File: cmd/podman/containers/ps.go [hunk 11]
```diff
@@ -93,6 +94,7 @@ func listFlagSet(cmd *cobra.Command) {
 	flags.Bool("noheading", false, "Do not print headers")
 	flags.BoolVarP(&listOpts.Size, "size", "s", false, "Display the total file sizes")
 	flags.BoolVar(&listOpts.Sync, "sync", false, "Sync container state with OCI runtime")
+	flags.BoolVar(&listOpts.Restart, "restart", false, "Display the container restart count")
```
Comment (Luap99) [inline cmd/podman/containers/ps.go:97]: I really do not like these flags, we already offer a complete customizable format with `--format` so I do not want to add flags for each option. This starts to add up quickly.

File: cmd/podman/containers/create.go (hunk 2 above)
Comment (vrothberg) [reply cmd/podman/containers/create.go:142]: Yes, that's still a breaking change.  Existing deployments using `--pod new:foo --restart=...` will start to break.

We can follow K8s _by default_ but that doesn't mean we need to restrict users from changing individual policies, especially if it worked before.  Don't we already support individual restart policies when configured in K8s YAML?

I am sure, we can find a solution where containers will _inherit_ the restart policy from the pod unless the user explicitly overwrites it.

File: cmd/podman/containers/ps.go (hunk 11 above)
Comment (vrothberg) [reply cmd/podman/containers/ps.go:97]: I share the feeling, @Luap99.  In fact, I do not see any reason to add this kind of information to `ps` at all.  The `inspect` commands in combination with `--format` should be used for that.

See https://github.com/containers/podman/pull/17830/files#diff-a06598ad58c5be472e9e040f89c040b4548837ce23c60fdb4eba27ba766c571fR189-R190 as an example.

File: cmd/podman/containers/create.go (hunk 2 above)
Comment (umohnani8) [reply cmd/podman/containers/create.go:142]: So just thinking a step further for `podman kube generate`, if the pod has a restart policy set to `on failure` and there are two containers in the pod where one has restart policy set to `never` and the other has no restart policy set, do we inherit the restart policy of the pod, which would be `on failure`.

During the kube generate command, do we just ignore the individual container's restart policy and set the policy for the whole pod to `on failure`? 

and what about when the pod has no restart policy set and only one container in it has it set to `never`, do we apply a restart policy of `never` to the whole pod then when doing kube generate?

I understand that this is a breaking change, but being able to have different restart policies for a pod and for a container within the pod seems confusing, and may be confusing for users who are probably coming from the k8s world or for users who use pods in podman and when they transition to k8s, they notice the behavior is quite different there.

We can ensure that deployments that use `--pod=new:foo --restart` don't break and only restrict the use of `--restart` with `podman create/run` when the `--pod` flag doesn't have a `new:` prefix.

File: docs/source/markdown/podman-pod-create.1.md.in (hunk 6 above)
Comment (umohnani8) [reply docs/source/markdown/podman-pod-create.1.md.in:149]: @vrothberg does this make sense or should `exit-policy` not be changed based on the pod's `restart policy`?

File: cmd/podman/containers/create.go (hunk 2 above)
Comment (vrothberg) [reply cmd/podman/containers/create.go:142]: > During the kube generate command, do we just ignore the individual container's restart policy and set the policy for the whole pod to on failure?

`generate kube` should reflect the restart policy of each container. If the pod has a restart policy set, this policy should bubble down to all containers by default unless explicitly overridden by the user.  So all containers will have policy set that `generate kube` can use.  If a container has no policy set, then its pod has none either.

> and what about when the pod has no restart policy set and only one container in it has it set to never, do we apply a restart policy of never to the whole pod then when doing kube generate?

An individual policy should only apply to the individual container.

> I understand that this is a breaking change, but being able to have different restart policies for a pod and for a container within the pod seems confusing, and may be confusing for users who are probably coming from the k8s world or for users who use pods in podman and when they transition to k8s, they notice the behavior is quite different there.

If users transition from K8s to Podman they will be using K8s YAML where we already allow to set individual restart policies for containers.

File: cmd/podman/containers/ps.go (hunk 3 above)
Comment (vrothberg) [reply cmd/podman/containers/ps.go:202]: Breaking it out into another PR or commit would be great, thanks!

File: cmd/podman/containers/ps.go (hunk 4 above)
Comment (vrothberg) [reply cmd/podman/containers/ps.go]: > Adding restarts to podman ps is part of the design doc goals

Apologies, I did not see that.

> It also helps with testing the pod restart policy feature.

The `inspect` commands already support showing the restart counts (which is used in other tests), so I don't think it helps with testing.

I don't feel extremely strongly about (not) adding the flag but if the motivation was to simplify testing, then we can avoid the extra code by just using `inspect`.

File: docs/source/markdown/podman-pod-create.1.md.in (hunk 6 above)
Comment (vrothberg) [reply docs/source/markdown/podman-pod-create.1.md.in:149]: Can we break this out into another issue/card?  Once we have the `--restart` behavior nailed down, I hope it would be easier to put the puzzle together.

File: cmd/podman/containers/create.go (hunk 2 above)
Comment (umohnani8) [reply cmd/podman/containers/create.go:142]: In the k8s yaml, you can only set the restart policy at the pod level, which is applied to all the containers in the pod. There is no way to specify a restart policy for each container in a pod in a k8s yaml. When we do a `kube generate`, there will only be one restart policy value set in the yaml at the pod level. 

That is why I was trying to understand/figure out that if we allow setting the restart policy for a pod and allow setting separate restart policies for containers within a pod, how do we determine what the final restart policy would be set at the pod level when doing `kube generate`.

I do think having restart policy behave differently for a pod in k8s and in podman will be confusing to the user and this is more of a missing feature or bug instead of breaking change if we decide to only allow setting the restart policy at the pod level when creating containers in a pod.

If we continue to allow setting separate restart policy for containers in a pod, I think the yaml generated via `kube yaml` should always just pick the pod restart policy - that would be `Always` by default or what the user set it to during pod creation.

cc @mheon @baude @rhatdan

Comment (vrothberg) [reply cmd/podman/containers/create.go:142]: > There is no way to specify a restart policy for each container in a pod in a k8s yaml. When we do a kube generate, there will only be one restart policy value set in the yaml at the pod level.

You are right, thanks for clarifying.

> I do think having restart policy behave differently for a pod in k8s and in podman will be confusing to the user and this is more of a missing feature or bug instead of breaking change if we decide to only allow setting the restart policy at the pod level when creating containers in a pod.

We must take backwards compatibility seriously.  It has been working so far and an unknown amount of existing workloads may break. K8s compat is important for `podman kube *` but, for me, it's not a strong argument to break `podman pod` behavior.  If it would be my workload, I'd be upset.

`podman pod` existed before `podman kube` and it's commonly used to bundle workloads together (e.g., easy communication via `localhost`).  IMO it is fundamentally different to K8s.

> If we continue to allow setting separate restart policy for containers in a pod, I think the yaml generated via kube yaml should always just pick the pod restart policy - that would be Always by default or what the user set it to during pod creation.

It's definitely something to be improved.  Looking at the current behavior, it seems like the last container's restart policy is set for the entire pod which seems extremely random.  Using the one of the pod seems like a good idea.

Comment (umohnani8) [reply cmd/podman/containers/create.go]: I agree, we don't want to break anyone. I have removed the restriction and added a note that in podman v5.0, we should reconsider this.

This PR focuses on adding restart policy to podman pod only, so I will open a follow up PR once this is in to make the changes for kube gen and play.

Thank you for the reviews!

File: docs/source/markdown/podman-pod-create.1.md.in (hunk 6 above)
Comment (umohnani8) [reply docs/source/markdown/podman-pod-create.1.md.in]: Removed it and will add in another PR

File: cmd/podman/containers/ps.go (hunk 4 above)
Comment (umohnani8) [reply cmd/podman/containers/ps.go]: I think it is still ideal for users to be able to get this info from ps, so I have left the option available for --format. I have removed the `--restart` flag for both `podman ps` and `podman pod ps`.

File: cmd/podman/containers/ps.go (hunk 11 above)
Comment (umohnani8) [reply cmd/podman/containers/ps.go]: Removed the flag.

File: cmd/podman/containers/ps.go (hunk 3 above)
Comment (umohnani8) [reply cmd/podman/containers/ps.go]: Removed and will open another PR to fix.

**Comment by umohnani8** [comment]
Once this is in, I will create a follow up PR for kube gen and play to start using the pod restart policy.

**Comment by umohnani8** [comment]
@edsantiago @mheon is there a way for me to not list `.RestartCount` in the docs? I want the option the users know and use to be `.Restarts`. 
This is the validation error I am getting https://github.com/containers/podman/pull/17627/checks?check_run_id=12662786279

**Comment by edsantiago** [comment]
/me groans in agony.

@Luap99 you might be able to answer that more concisely and accurately than anything I could do.

**Comment by umohnani8** [comment]
Tests are green, @containers/podman-maintainers PTAL

File: cmd/podman/containers/create.go [hunk 12]
```diff
@@ -137,6 +137,7 @@ func create(cmd *cobra.Command, args []string) error {
 		}
 		cliVals.InitContainerType = initctr
 	}
+	// TODO: v5.0 block users from setting restart policy for a container if the container is in a pod
```
Comment (vrothberg) [inline cmd/podman/containers/create.go:140]: I am still strongly objecting to that but we can wrestle that out when time is due :^)

File: docs/source/markdown/podman-pod-clone.1.md.in [hunk 13]
```diff
@@ -67,6 +67,10 @@ Set a custom name for the cloned pod. The default if not specified is of the syn
 
 @@option pid.pod
 
+@@option restart
+
+The restart policy is set at the pod level so all containers in the pod have the same restart policy configured.
```
Comment (vrothberg) [inline docs/source/markdown/podman-pod-clone.1.md.in:72]: The description isn't accurate anymore since containers can still use a different policy if the user wishes to.  How about:
`Default restart policy for all container in the pod.`?

File: docs/source/markdown/podman-pod-create.1.md.in [hunk 14]
```diff
@@ -143,6 +143,10 @@ but only by the pod itself.
 
 @@option replace
 
+@@option restart
+
+The restart policy is set at the pod level so all containers in the pod have the same restart policy configured.
```
Comment (vrothberg) [inline docs/source/markdown/podman-pod-create.1.md.in:148]: Same here.

File: pkg/specgen/generate/container_create.go [hunk 15]
```diff
@@ -555,12 +555,27 @@ func createContainerOptions(rt *libpod.Runtime, s *specgen.SpecGenerator, pod *l
...
 	// Default used if not overridden on command line
 
-	if s.RestartPolicy != "" {
+	var (
+		restartPolicy string
+		retries       uint
+	)
+	// If the container is running in a pod, use the pod's restart policy for all the containers
+	if pod != nil {
+		podConfig, err := pod.Config()
```
Comment (vrothberg) [inline pkg/specgen/generate/container_create.go:564]: Just noticed: `pod.Config()` is a very expensive call as it's re-marshaling the entire config of the pod.

Non-blocking for this PR but please address it in a future one due avoid the perf regression: we should add `(*Pod).ConfigNoCopy()` as we did for containers.

**Review by vrothberg — COMMENTED** [review]
LGTM other than the final nits

File: docs/source/markdown/podman-pod-clone.1.md.in (hunk 13 above)
Comment (umohnani8) [reply docs/source/markdown/podman-pod-clone.1.md.in]: Fixed

File: docs/source/markdown/podman-pod-create.1.md.in (hunk 14 above)
Comment (umohnani8) [reply docs/source/markdown/podman-pod-create.1.md.in]: Fixed

**Comment by vrothberg** [comment]
@mheon, no need to backport https://github.com/containers/podman/pull/18481 since the changes here did not make it into 4.5.  The reports were against 4.6-dev, so we're good.
